projects/core/src/cart/facade/cart-voucher.service.ts

```diff
@@ -0,0 +1,93 @@
+import { Injectable } from '@angular/core';
+import { select, Store } from '@ngrx/store';
+import { Observable } from 'rxjs';
+import { filter } from 'rxjs/operators';
+import { AuthService } from '../../auth/index';
+import { Cart } from '../../model/cart.model';
+import { OCC_USER_ID_ANONYMOUS } from '../../occ/utils/occ-constants';
+import * as fromProcessStore from '../../process/store/process-state';
+import {
+  getProcessErrorFactory,
+  getProcessLoadingFactory,
+  getProcessSuccessFactory,
+} from '../../process/store/selectors/process.selectors';
+import { CartActions } from '../store/actions/index';
+import { ADD_VOUCHER_PROCESS_ID, StateWithCart } from '../store/cart-state';
+import { CartSelectors } from '../store/selectors/index';
+
+@Injectable()
+export class CartVoucherService {
+  private _userId = OCC_USER_ID_ANONYMOUS;
+  private _cart: Cart;
+
+  constructor(
+    protected store: Store<
+      StateWithCart | fromProcessStore.StateWithProcess<void>
+    >,
+    protected authService: AuthService
+  ) {
+    this.authService
```

Comment:
Can we make it reactive,i.e. not subscribe and listen all the time, but only get the data when needed? 
For example:
```
  private readonly userId$ = this.authService.getUserToken().pipe(
    map(userToken => {
      if (Object.keys(userToken).length !== 0) {
        return userToken.userId;
      } else {
        return OCC_USER_ID_ANONYMOUS;
      }
    })
  );

  private readonly cartId$: Observable<string> = this.store.pipe(
    select(CartSelectors.getCartContent),
    map(cart => cart.code)
  );
```

and 

```
  addVoucher(voucherId: string): void {
    combineLatest([this.userId$, this.cartId$])
      .pipe(take(1))
      .subscribe(([userId, cartId]) => {
        this.store.dispatch(
          new CartActions.CartAddVoucher({
            userId: userId,
            cartId: cartId,
            voucherId: voucherId,
          })
        );
      });
  }

```

Reply:
Done, not completely follow your way, but I get the both when needed